# 04. Embeddings e indexacion en PostgreSQL + pgvector

Este notebook implementa la carga del corpus RAG en PostgreSQL, la generacion de embeddings y la validacion de persistencia en `pgvector` usando la capa productiva de `src.vector_store`.

> Si cambias `EMBEDDING_MODEL`, reejecuta el notebook completo para regenerar los embeddings persistidos en PostgreSQL.

## Objetivos

- cargar el dataset `rag.parquet`
- crear o validar el esquema `pgvector`
- insertar convocatorias por lotes usando `src.vector_store`
- comprobar el numero de registros persistidos
- validar cobertura minima del indice vectorial y consistencia de metadatos


In [1]:
from __future__ import annotations

import sys
from pathlib import Path

ROOT = Path.cwd().resolve().parent if Path.cwd().name == "notebooks" else Path.cwd().resolve()
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

import pandas as pd
from IPython.display import display
from sqlalchemy import text
from tqdm import tqdm

from src.config import PRIMARY_RAG_PATH, RAG_TEXT_COLUMN, settings
from src.data_loader import load_retrieval_corpus
from src.vector_store import connect_db, create_schema_if_needed, insert_convocatorias

BATCH_SIZE = 64
SAMPLE_LIMIT: int | None = None

print("ROOT:", ROOT)
print("DATABASE_URL:", settings.database_url)
print("EMBEDDING_MODEL:", settings.embedding_model)
print("RAG_PATH:", PRIMARY_RAG_PATH)


ROOT: /var/www/codigo/maestria_ia/umsa/diplomados_intermedios/dip_03
DATABASE_URL: postgresql+psycopg://postgres:postgres@localhost:5432/sicoes_rag
EMBEDDING_MODEL: sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2
RAG_PATH: /var/www/codigo/maestria_ia/umsa/diplomados_intermedios/dip_03/data/rag/sicoes_convocatorias_rag.parquet


## 1. Cargar corpus RAG

El corpus indexado parte del dataset RAG canónico versionado en `parquet`.


In [2]:
df = load_retrieval_corpus()
if SAMPLE_LIMIT is not None:
    df = df.head(SAMPLE_LIMIT).copy()

print("Filas cargadas:", len(df))
print("Columnas:", df.columns.tolist())
df[["document_id", "cuce", "entidad", RAG_TEXT_COLUMN]].head(3)


Filas cargadas: 1429
Columnas: ['document_id', 'cuce', 'entidad', 'tipo_contratacion', 'modalidad', 'objeto_contratacion', 'subasta', 'fecha_publicacion', 'fecha_presentacion', 'fecha_publicacion_iso', 'fecha_presentacion_iso', 'estado', 'archivos_disponibles', 'ficha_url', 'source_draw', 'source_row_in_page', 'longitud_objeto', 'cantidad_palabras_objeto', 'texto_rag']


,document_id,cuce,entidad,texto_rag
0,26-1101-00-1669685-1-1,26-1101-00-1669685-1-1,Gobierno Autonomo Municipal De Sucre,CUCE: 26-1101-00-1669685-1-1\nEntidad: Gobiern...
1,26-0417-03-1669697-1-1,26-0417-03-1669697-1-1,Caja Nacional De Salud Regional Santa Cruz,CUCE: 26-0417-03-1669697-1-1\nEntidad: Caja Na...
2,26-0417-03-1669695-1-1,26-0417-03-1669695-1-1,Caja Nacional De Salud Regional Santa Cruz,CUCE: 26-0417-03-1669695-1-1\nEntidad: Caja Na...


## 2. Crear esquema y validar conexion

Esta celda usa el SQL de inicializacion y la configuracion de `src.vector_store`. Si falla en tu entorno, revisa `DATABASE_URL` y el estado del contenedor Docker.


In [3]:
create_schema_if_needed()
engine = connect_db()

with engine.connect() as connection:
    print("SELECT 1 ->", connection.exec_driver_sql("SELECT 1").scalar())
    table_exists = connection.exec_driver_sql(
        "SELECT to_regclass('public.convocatorias')"
    ).scalar()
    print("Tabla convocatorias:", table_exists)


SELECT 1 -> 1
Tabla convocatorias: convocatorias


## 3. Insercion por lotes

La insercion se hace mediante `insert_convocatorias()` con `upsert`, por lo que el notebook puede reejecutarse sin duplicar registros por `document_id`.


In [4]:
records = df.to_dict(orient="records")
total_batches = (len(records) + BATCH_SIZE - 1) // BATCH_SIZE
inserted_total = 0

for batch_index in tqdm(range(total_batches), total=total_batches, desc="Insertando lotes", unit="lote"):
    start = batch_index * BATCH_SIZE
    end = start + BATCH_SIZE
    batch = records[start:end]
    inserted = insert_convocatorias(batch, generate_missing_embeddings=True, upsert=True)
    inserted_total += inserted

print("Total de registros procesados:", inserted_total)
print("Total de lotes:", total_batches)


Insertando lotes: 100%|██████████| 23/23 [00:42<00:00,  1.84s/lote]

Total de registros procesados: 1429
Total de lotes: 23


## 4. Validaciones posteriores a la insercion


In [5]:
with engine.connect() as connection:
    total_rows = connection.execute(text("SELECT COUNT(*) FROM convocatorias")).scalar_one()
    rows_with_embeddings = connection.execute(
        text("SELECT COUNT(*) FROM convocatorias WHERE embedding IS NOT NULL")
    ).scalar_one()
    sample_rows = pd.read_sql(
        text(
            """
            SELECT document_id, cuce, entidad, modalidad, estado
            FROM convocatorias
            ORDER BY id
            LIMIT 5
            """
        ),
        connection,
    )

print("Filas totales en PostgreSQL:", total_rows)
print("Filas con embedding:", rows_with_embeddings)
sample_rows


Filas totales en PostgreSQL: 1429
Filas con embedding: 1429


,document_id,cuce,entidad,modalidad,estado
0,26-1101-00-1669685-1-1,26-1101-00-1669685-1-1,Gobierno Autonomo Municipal De Sucre,ANPP,Vigente
1,26-0417-03-1669697-1-1,26-0417-03-1669697-1-1,Caja Nacional De Salud Regional Santa Cruz,CNC,Vigente
2,26-0417-03-1669695-1-1,26-0417-03-1669695-1-1,Caja Nacional De Salud Regional Santa Cruz,CNC,Vigente
3,26-0418-07-1669627-1-1,26-0418-07-1669627-1-1,Caja Petrolera De Salud Administracion Departa...,ANPP,Vigente
4,26-1205-00-1647943-2-1,26-1205-00-1647943-2-1,Gobierno Autonomo Municipal De El Alto,ANPE,Vigente


## 5. Validacion estructural posterior a la indexacion

Esta seccion evita mezclar ingestion con evaluacion de retrieval. Aqui solo se verifican cobertura de embeddings, unicidad y una distribucion basica de metadatos persistidos.


In [6]:
with engine.connect() as connection:
    validation_summary = pd.read_sql(
        text(
            """
            SELECT
                COUNT(*) AS total_rows,
                COUNT(DISTINCT document_id) AS unique_document_ids,
                COUNT(*) FILTER (WHERE embedding IS NOT NULL) AS rows_with_embedding,
                COUNT(*) FILTER (WHERE metadata_json IS NOT NULL) AS rows_with_metadata,
                ROUND(AVG(LENGTH(texto_rag))::numeric, 2) AS avg_texto_rag_len
            FROM convocatorias
            """
        ),
        connection,
    )
    modalidad_distribution = pd.read_sql(
        text(
            """
            SELECT modalidad, COUNT(*) AS total
            FROM convocatorias
            GROUP BY modalidad
            ORDER BY total DESC, modalidad ASC
            LIMIT 10
            """
        ),
        connection,
    )

display(validation_summary)
display(modalidad_distribution)


,total_rows,unique_document_ids,rows_with_embedding,rows_with_metadata,avg_texto_rag_len
0,1429,1429,1429,1429,422.48


,modalidad,total
0,ANPP,460
1,ANPE,433
2,CM,188
3,CNC,111
4,LP,94
5,CND1,75
6,OF,67
7,CD,1


## 6. Chequeos rapidos de consistencia


In [7]:
with engine.connect() as connection:
    consistency_checks = pd.read_sql(
        text(
            """
            SELECT
                COUNT(*) FILTER (WHERE cuce IS NULL OR cuce = '') AS missing_cuce,
                COUNT(*) FILTER (WHERE entidad IS NULL OR entidad = '') AS missing_entidad,
                COUNT(*) FILTER (WHERE objeto_contratacion IS NULL OR objeto_contratacion = '') AS missing_objeto,
                COUNT(*) FILTER (WHERE texto_rag IS NULL OR texto_rag = '') AS missing_texto_rag,
                COUNT(*) - COUNT(DISTINCT cuce) AS duplicate_cuce_gap
            FROM convocatorias
            """
        ),
        connection,
    )
    estado_distribution = pd.read_sql(
        text(
            """
            SELECT estado, COUNT(*) AS total
            FROM convocatorias
            GROUP BY estado
            ORDER BY total DESC, estado ASC
            """
        ),
        connection,
    )

display(consistency_checks)
display(estado_distribution)


,missing_cuce,missing_entidad,missing_objeto,missing_texto_rag,duplicate_cuce_gap
0,1,0,0,0,1


,estado,total
0,Vigente,1429


## 7. Observaciones

- Este notebook reutiliza la capa `src.vector_store` para evitar duplicacion de logica.
- La evaluacion comparativa de retrieval (`keyword`, `semantic`, `hybrid`) se realiza en `05_rag_evaluation.ipynb`.
- Si la conexion local falla, verificar Docker, `DATABASE_URL` y que el puerto `5432` este accesible desde el entorno Python usado por el notebook.
